src/python/pants/backend/experimental/python/lockfile_metadata.py

```diff
@@ -20,15 +19,24 @@ class LockfileMetadata:
     requirements_invalidation_digest: str | None
     valid_for_interpreter_constraints: InterpreterConstraints | None
 
-    @staticmethod
-    def from_json_bytes(json_literal: str | bytes) -> LockfileMetadata:
-        """Reads a `string`/`bytes` that has JSON-encoded a `LockfileMetadata` object, returning
-        that deserialized object."""
+    @classmethod
+    def from_lockfile(cls, lockfile: bytes) -> LockfileMetadata:
+        """Parse all relevant metadata from the lockfile's header."""
+        in_metadata_block = False
+        metadata_lines = []
+        for line in lockfile.splitlines():
+            if line == BEGIN_LOCKFILE_HEADER:
+                in_metadata_block = True
+            elif line == END_LOCKFILE_HEADER:
+                break
+            elif in_metadata_block:
+                metadata_lines.append(line[2:])
+
         try:
-            metadata = json.loads(json_literal)
+            metadata = json.loads(b"\n".join(metadata_lines))
         except json.decoder.JSONDecodeError:
-            # If this block is invalid, this should trigger error/warning behavior
-            metadata = {}
+            # TODO(#12314): Add a good error.
+            raise
```

Comment:
Note that on line 49 here, there's another TODO where we were leaving error handling to the calling context. It's probably worth figuring out what the correct behaviour is for both of these.

(see https://github.com/pantsbuild/pants/pull/12611/files#diff-cdc1b4b3d027d106c4bbba50a949d41e9fc18c3990c0115c9402b559785bdcfdR46-R50)